### PR TITLE
SDI-34 get visit ids by filter endpoint

### PR DESCRIPTION
**Pull request by lbennett-moj**
Using jpa specification

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Address.kt [hunk 1]
```diff
@@ -31,7 +31,7 @@ abstract class Address {
   @SequenceGenerator(name = "ADDRESS_ID", sequenceName = "ADDRESS_ID", allocationSize = 1)
   @GeneratedValue(generator = "ADDRESS_ID")
   @Column(name = "ADDRESS_ID", nullable = false)
-  val addressId: Long = 0
+  open val addressId: Long = 0
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Address.kt:35]: to remove hibernate errors in logs

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Visit.kt [hunk 2]
```diff
@@ -119,6 +119,9 @@ data class Visit(
   /* a list of all visitors including those without visitor orders */
   @OneToMany(mappedBy = "visit", cascade = [CascadeType.ALL])
   val visitors: MutableList<VisitVisitor> = mutableListOf(),
+
+  @Column(name = "CREATE_DATETIME", nullable = false, insertable = false, updatable = false)
+  var whenCreated: LocalDateTime = LocalDateTime.now()
 ) {
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/Visit.kt:125]: required for querying visits between date ranges during nomis migration

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt [hunk 3]
```diff
@@ -104,4 +108,10 @@ class Repository(
     val visit = visitRepository.findById(visitId!!).orElseThrow()
     visit.visitStatus = visitStatusRepository.findById(VisitStatus.NORM).orElseThrow()
   }
+
+  fun updateCreatedToMatchVisitStart() {
+    val sql =
+      "UPDATE offender_visits SET CREATE_DATETIME = START_TIME"
+    jdbcTemplate!!.execute(sql)
+  }
```
Comment (lbennett-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt:116]: the CREATE_DATETIME is a DB system timestamp on creation. Setting the CREATE_DATETIME to match the visit date makes the integration tests testing date ranges clearer

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt [hunk 4]
```diff
@@ -32,6 +34,8 @@ class Repository(
   val visitStatusRepository: ReferenceCodeRepository<VisitStatus>,
   val visitTypeRepository: ReferenceCodeRepository<VisitType>,
 ) {
+  @Autowired
+  var jdbcTemplate: JdbcTemplate? = null
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt:38]: you could do 
`var lateinit jdbcTemplatel/ JdbcTemplate`   to get rid of the nullable thing

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt [hunk 5]
```diff
@@ -104,4 +108,10 @@ class Repository(
     val visit = visitRepository.findById(visitId!!).orElseThrow()
     visit.visitStatus = visitStatusRepository.findById(VisitStatus.NORM).orElseThrow()
   }
+
+  fun updateCreatedToMatchVisitStart() {
+    val sql =
+      "UPDATE offender_visits SET CREATE_DATETIME = START_TIME"
+    jdbcTemplate!!.execute(sql)
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/Repository.kt:115]: see above on `!!`

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**